pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -681,5 +862,66 @@ void topicTerminated() {
         }
     }
 
+    /**
+     * Commit a transaction.
+     *
+     * @param txnID         {@link TxnID} to identify the transaction.
+     * @param properties    Additional user-defined properties that can be associated with a particular cursor position.
+     * @throws IllegalArgumentException
+     * @throws TransactionConflictException
+     */
+    public synchronized void commitTxn(TxnID txnID, Map<String,Long> properties) throws TransactionConflictException {
+        if (!this.pendingAckMessagesMap.containsKey(txnID)) {
```

Comment:
Actually this check can be moved down to line 890. so we just do `remove` and check if the TxnID can be be removed. so it is one operation to prevent any concurrent commits on same TxnID.

```
 ConcurrentOpenHashSet<Position> pendingAckMessageCurTxn = this.pendingAckMessageMap.remove(txnID);
if (null == pendingAckMessageCurTxn) {
     throw new TransactionNotFoundException(...);
} 
```

Reply:
make sense, will update.